rtrlib/pfx/trie/trie-pfx.c

```diff
@@ -0,0 +1,472 @@
+                    pfx_table_notify_clients(pfx_table, &record, false);
+                }
+                rm_node = (trie_remove(root, &(root->prefix), root->len, 0));
+                assert(rm_node != NULL);
+                free(((struct node_data *) rm_node->data)->ary);
+                free(rm_node->data);
+                free(rm_node);
+            } while(rm_node != root);
+            if(i == 0)
+                pfx_table->ipv4 = NULL;
+            else
+                pfx_table->ipv6 = NULL;
+
+            pthread_rwlock_unlock(&(pfx_table->lock));
+        }
+    }
+    pthread_rwlock_destroy(&(pfx_table->lock));
+}
+
+int pfx_table_append_elem(struct node_data *data, const struct pfx_record *record)
+{
+    data->len++;
+    data->ary = realloc(data->ary, sizeof(struct data_elem) * data->len);
+    if(data->ary == NULL)
+        return PFX_ERROR;
+    data->ary[data->len - 1].asn = record->asn;
+    data->ary[data->len - 1].max_len = record->max_len;
+    data->ary[data->len - 1].socket = record->socket;
+    return PFX_SUCCESS;
+}
+
+int pfx_table_create_node(struct trie_node **node, const struct pfx_record *record)
+{
+    *node = malloc(sizeof(struct trie_node));
+    if(*node == NULL)
+        return PFX_ERROR;
+    (*node)->prefix = record->prefix;
+    (*node)->len = record->min_len;
+    (*node)->lchild = NULL;
+    (*node)->rchild = NULL;
+    (*node)->parent = NULL;
+
+    (*node)->data = malloc(sizeof(struct node_data));
+    if((*node)->data == NULL)
+        return PFX_ERROR;
+    ((struct node_data *) (*node)->data)->len = 0;
+    ((struct node_data *) (*node)->data)->ary = NULL;
+
+    return pfx_table_append_elem(((struct node_data *) (*node)->data), record);
+}
+
+struct data_elem *pfx_table_find_elem(const struct node_data *data, const struct pfx_record *record, unsigned int *index)
+{
+    for(unsigned int i = 0; i < data->len; i++) {
+        if(data->ary[i].asn == record->asn && data->ary[i].max_len == record->max_len && data->ary[i].socket == record->socket) {
+            if(index != NULL)
+                *index = i;
+            return &(data->ary[i]);
+        }
+    }
+    return NULL;
+}
+
+//returns pfx_table->ipv4 if record version is LRTR_IPV4 else pfx_table->ipv6
+inline struct trie_node *pfx_table_get_root(const struct pfx_table *pfx_table, const enum lrtr_ip_version ver)
+{
+    return ver == LRTR_IPV4 ? pfx_table->ipv4 : pfx_table->ipv6;
+}
+
+int pfx_table_del_elem(struct node_data *data, const unsigned int index)
+{
+    //if index is not the last elem in the list, move all other elems backwards in the array
+    if(index != data->len - 1) {
+        for(unsigned int i = index; i < data->len - 1; i++) {
+            data->ary[i] = data->ary[i+1];
+        }
+    }
+    data->len--;
+    data->ary = realloc(data->ary, sizeof(struct data_elem) * data->len);
+    if(data->len != 0 && data->ary == NULL)
```

Comment:
This seems a little bit difficult.
@reuteran fix was:
```
@@ -141,8 +144,10 @@ int pfx_table_del_elem(struct node_data *data, const unsigned int index)
     }
     data->len--;
     data->ary = realloc(data->ary, sizeof(struct data_elem) * data->len);
-    if(data->len != 0 && data->ary == NULL)
+    if(data->len != 0 && data->ary == NULL) {
+        free(&data->ary[data->len]);
         return PFX_ERROR;
+    }
     else if(data->len == 0)
         data->ary = NULL;
```

The free() manpage says:

> The  free()  function frees the memory space pointed to by ptr, which must have been returned by a previous call
>        to malloc(), calloc(), or realloc().
> 

free() would be called not for the starting address of the allocated memory area. If I understand the manpage right it's a requirement that the exact passed address was previously returned by a memory allocation function.
So I think free() would fail here, maybe cause a segfault.

If realloc() fails here we would have the state that the array is 1 element bigger than we need and the unused element would be at the end.
The data->len would contain the new length already.
So to handle this situation correctly we could  restore the previous state.
This means shift the elements in the array again and insert the old value again.
I don't know if this makes so much sense.
We can also keep the current state and, the array would be 1 element bigger than needed in the very unlikely case that realloc() fails when we want to decrease the allocated memory size (not sure if this even can happen).
When the next element would be appened or deleted from the array another remalloc() try would be made. If that call would be succesful it would also fix the previous failure.

If data->len becomes 0 and we have X unused elements in the array I would expect that remalloc() can't fail because it would be the same than a call to free().

What do you think @nbars  @reuteran  @smlng?